### ISSUE TITLE
Iterable.none expects at least one element

### ISSUE THREAD

**Issue by jGleitz** [robstoll/atrium#689]
*Affected Version*: 0.13.0
*API* (fluent-en_GB, infix-en_GB): fluent-en_GB
*Platform* (jvm, js): jvm
*Extension* (none, kotlin 1.3): none

## How to reproduce the problem  
Consider the following code snippet
```kotlin
expect(emptyList<Int>()).none { isGreaterThan(6) }
```

## Describe the bug
The assertion leads to the following output:
```
expected that subject: []        (kotlin.collections.EmptyList <2051387231>)
◆ does not contain: 
  ⚬ an entry which: 
      » is greater than: 6        (kotlin.Int <1355096431>)
    ✔ ▶ number of such entries: 0
        ◾ is: 0        (kotlin.Int <1212607454>)
    ✘ ▶ has at least one element: false
        ◾ is: true
```

## Expected behaviour
The assertion should hold. To me, “none” pretty clearly means “for all: not”, and as we all know, all-quantified statements about the empty set are always true. In short: `none` should not require at least one element (neither should the `containsNot` build in general, if you ask me).

**Comment by robstoll** [comment]
It's not a bug, it's a feature :wink: 

You are right from a theoretical stand point but in practice, in most cases I have seen so far, it is a crucial difference if an Iterable has no next element or if it has an arbitrary number of elements and they all need to fulfil certain criteria. I have seen enough bugs which were there because one forgot the empty case (especially with `all`). Atrium helps to prevent such mistakes. Especially in a test one should use `isEmpty()` to be clear about the outcome. The check for at least one element also prevents defective test setups where for instance an Iterable gets consumed before using `none` or `all`. Without this check it would be hidden. 

Maybe you can share your real use case so that I better understand it and maybe we can still improve Atrium somehow to better fulfil your needs.

**Comment by jGleitz** [comment]
My use case is this: I have a collection of task dependencies and want to make sure that while dependencies were added to some tasks, they were not added to others. Because the tasks also have other dependencies, I don’t want to do an exact match on the set of dependencies. This would break needlessly whenever the dependencies change for some unrelated reason.

Because of that, I use `none` because it (theoretically) expresses what I want to achieve: It checks that the dependencies were not added, but ignores all other dependencies. But because `none` does not accept no dependencies, this approach does not work currently. Some tasks have no dependencies, but I do not want to put this knowledge in my tests, because it would be a maintenance burden.

<hr>

I share your experience that developers sometimes fail to consider the case of empty collections. You are right that that is a common source for mistakes, and you are right that it would be great if atrium could help to avoid these mistakes.

*However*, I disagree that this consideration warrants breaking the principle of least surprise. As I have said before, I think it should be as easy as possible to understand test logic by just reading it in one pass. `all` and `none` are well-known words/operators with well-known semantics. From my point of view, even the best-intentioned efforts to avoid bugs do *not* justify breaking conventions and thus violating users’ justified expectations.

From my point of view, it would make sense to have `isNotEmptyAndNone` and `isNotEmptyAndAll` as shortcuts and to promote those very clearly in the docs of `all` and `none`.

**Comment by robstoll** [comment]
Thanks for sharing your use case. I am still not convinced to change the behaviour of none and all: people will not read the docs and just use `all` neglecting the empty case. Hence, I prefer if a more experienced dev stumbles over `none` and `all` as you did and would need to use `isEmptyOrNone` than the opposite. Reading `isEmptyOrNone` is as easy to read in one pass as `none` and states the implicit knowledge explicit. So... yes, I am with you that we need to document this in `none` and `all` as well as in `isEmptyOrNone` and `isEmptyOrAll` and I am willing to introduce isEmptyOr... but kind of want to have save defaults.
Good compromise?

**Comment by jGleitz** [comment]
> Good compromise?

Unfortunately, no. My main issue is *not* that I have to use a longer/more complicated assertion. My issue is that the code does not behave like any reasonable reader would expect it to behave.

I completely get your point that people will not necessarily read the docs. I don’t even want to dispute that if we changed `none` and `all`, some developers might possibly create more bugs in their code than if we did not change it.

I just think that following the principle of least surprise and having a predictable API is more important. You value avoiding potential bugs over predictability, I value it the other way around. I think the code fulfilling reasonable expectations is very, very important. I can’t find the source, but some developer wrote somewhere something like: “If an API surprises me, I stop trusting my intuition and start reading the documentation and/or source code for everything. This makes me 3x slower”. I agree 100% and think that not being able to trust my intuition is a huge disadvantage for an API.

In other words: Harming the trust of the many for the carelesness of the few seems like the wrong trade-off to me.

**Comment by jGleitz** [comment]
Footnote: It might seem silly to argue about those two assertions. And I agree: For me, this discussion is less about these specific assertions and more about the general trade-offs and principles atrium is built on. And I am, of course, aware that you will always be the final judge of those decisions. I just try to make my case the best I can.

**Comment by robstoll** [comment]
>  I just try to make my case the best I can.

and that's more than welcome. Also, I am more than happy if you find fellow dev supporting your view it's not like we can change things in Atrium (I added the label `needs votes`. But for now... I would agree with you if we talked about an API used to write business code. But we talk about an API to write tests. It's not a problem if the assertion asserts more than you expected. I don't think that one loses the trust IMO but I think rather the opposite is the case: "WTF why did it fail... oh.. I see where this is going, Atrium tries to protect newbies, fair enough... " Might be you stumble over this more than once which might be a bit annoying but you get used to I guess. And in the worst case, you write your own `none` which calls isEmptyOrNone and use it instead.

After writing this, I think it would also make sense to add a note to the error message in case the Iterable was empty, something along the line of:
```
expected that subject: []        (kotlin.collections.EmptyList <2051387231>)
◆ does not contain: 
  ⚬ an entry which: 
      » is greater than: 6        (kotlin.Int <1355096431>)
    ✔ ▶ number of such entries: 0
        ◾ is: 0        (kotlin.Int <1212607454>)
    ✘ ▶ has at least one element: false
        ◾ is: true    
        ℹ️ use isEmptyOrNone if you don't care about the empty case
```
